### DISCUSSION TITLE
Essentials Discord additions for further configuration (PRs ready)

### DISCUSSION THREAD

**Discussion by x86-39** [EssentialsX/Essentials#5155] [Ideas and Feature Suggestions] [answered]
Hi, we are looking at moving to Essentials Discord from DiscordSRV.  

There's a few features we still need which Essentials Discord does not have.  

They are:  
- Ability to set an alias for Discord roles for in Minecraft.  
  Eg. the "Admins" role shows up as "Admin" in Minecraft.  
  Can also be used to hardcode rolour colours, for example "Admins" role shows up as "&9Admin" in Minecraft. This is `DiscordChatChannelRoleAliases` in DiscordSRV  
- Ability to ignore certain roles from Discord by only looking at select roles. We only want Minecraft related roles to go to the Minecraft server, so we want a list of roles it accepts. This is `DiscordChatChannelRolesSelection` in DiscordSRV.  
- Ability to separate what types of messages go to the player webhook. Currently if you enable webhooks with `show-avatars`, all messages will go through that. We want messages such as joins, leaves, player deaths and advancements to still be sent by the bot user.  
- Ability to format the webhooks username. We want to put the primary group of the player in front of the username, but Essentials Discord just has two booleans for whether to use the username or displayname.  
- A separate format for messages when a player joins for the first time  

All of these features I have developed on my branches of Essentials.

The [discord-additions](https://github.com/diademiemi/Essentials/tree/discord-additions) branch contains all of these features. There are also 4 split branches where these features are included individually with the exception of the first two being one branch, as they are from within the same function.  

I saw in CONTRIBUTING.md an issue should be opened first before creating a pull request. If these features are welcome in Essentials Discord I will gladly create PRs for these.  

I believe these features go a long way to making the plugin more configurable and would be a benefit to the plugin. 

However, it doesn't seem like Essentials Discord has automatic config updating. For most features I've left the defaults as the current behavior, so older configs should have no change.  
For the branch that adjusts the webhook name format, there is a legacy function to check the existing booleans for if the format is undefined.

**Comment by JRoy** [comment]
Feel free to open your PR and I'll have a better look!

**Comment by x86-39** [comment, marked as answer]
Sent them all!

Add roles selection option for Discord #5157  
Allow configuring which types of messages get sent as webhook for Discord #5158  
Allow configuring webhook name with greater control for Discord #5159  
Add first join message type for Discord #5160

**Comment by x86-39** [comment]
They're all merged on my branch `discord-additions` so they shouldn't have any merge conflicts
